NMS refactor

hgaiser opened the pull request:
This PR refactors NMS, which has a couple of effects:

- The number of outputs for RetinaNet and their meaning changes. Before there was one additional output, `nms_classification`, this changes to `nms_boxes`, `nms_scores`, `nms_labels`. `nms_boxes` is shaped `(300, 4)`, `nms_scores` is shaped `(300,)` and `nms_labels` is shaped `(300,)`. The boxes define the `(x1, y1, x2, y2)` of the detected objects, the scores the classification scores and the labels the predicted labels.
- NMS now supports batch_size > 1.
- The evaluation and visualization code changes accordingly.
- Patch update to the pretrained COCO weights to support the new NMS parameters.

hgaiser commented:
Oh also, it speeds up drawing of detections significantly. After https://github.com/fizyr/keras-retinanet/pull/368 the drawing takes ~20msec. With this PR it is ~1msec. It also fixes https://github.com/fizyr/keras-retinanet/issues/334 .

Ori226 commented:
Does this PR also solve #383 ?

hgaiser commented:
No but it does open the possibility to fix that (I'm working on it at the moment).

de-vri-es commented on keras_retinanet/utils/visualization.py:
```diff
@@ -59,25 +59,25 @@ def draw_boxes(image, boxes, color, thickness=2):
+        scores    : A list of N classification scores.
+        labels    : A list of N labels.
+        color     : The color of the boxes. By default the color from keras_retinanet.utils.colors.label_color will be used.
+        generator : (optional) Generator which can map label to class name.
```
`score_threshold` isn't documented

hgaiser replied:
Nice catch.

de-vri-es commented on keras_retinanet/layers/nms.py:
```diff
@@ -0,0 +1,189 @@
+    all_other  = []
+
+    # perform per class NMS
+    for c in range(int(classification.shape[1])):
```
Aren't shapes always ints anyway?

hgaiser replied:
They're actually tensors, but in this case they are always representing an int.

de-vri-es replied:
I did say int**s** (plural) ^^ But I'm wondering if the cast to `int` is really needed.

de-vri-es commented on keras_retinanet/layers/nms.py:
```diff
@@ -0,0 +1,189 @@
+    labels   = keras.backend.cast(labels, 'int32')
+    other    = [backend.pad(o, [[0, pad_size]] + [[0, 0] for _ in range(1, len(o.shape))], constant_values=-1) for o in other]
+
+    return [boxes, scores, labels] + other
```
Looks clean, right amount of comments too :+1:

hgaiser commented on keras_retinanet/layers/nms.py:
```diff
@@ -0,0 +1,189 @@
+        classification  : Tensor of shape (num_boxes, num_classes) containing the classification scores.
+        other           : List of tensors of shape (num_boxes, ...) to filter along with the boxes and classification scores.
+        score_threshold : Threshold used to prefilter the boxes with.
+        max_detections  : Maximum number of detections to filter.
```
Yep.

de-vri-es replied:
It still says to filter.

de-vri-es commented on keras_retinanet/bin/convert_model.py:
```diff
@@ -0,0 +1,60 @@
+    parser = argparse.ArgumentParser(description='Script for converting a training model to an inference model.')
+
+    parser.add_argument('model', help='The model to convert.')
+    parser.add_argument('converted_model_path', help='Path to save the converted model to.')
```
maybe just name the parameters `model_in` and `model_out` or something? Probably looks better in the help text.

hgaiser replied:
:+1:

hgaiser commented:
Updated the PR. The following changes have been made:

- Training works with training models. In other words, the layers required for inference are not added during training (only for evaluation purposes).
- A script is added to convert from a training model to an inference model (which always has the outputs `[boxes, scores, labels, other[0], other[1], ...]`.
- Renamed NonMaximumSuppression layer to FilterDetections layer (as it can be configured to skip NMS now).
- Moved some backbone specific stuff to `keras_retinanet/models/__init__.py`.

de-vri-es commented on keras_retinanet/bin/train.py:
```diff
@@ -78,19 +80,14 @@ def create_models(backbone_retinanet, backbone, num_classes, weights, multi_gpu=
+        training_model = model
+
+    # make prediction model
+    prediction_model = retinanet_bbox(model=model)
```
New model initialization looks good. A lot cleaner :)

de-vri-es commented on keras_retinanet/bin/train.py:
```diff
@@ -291,18 +288,7 @@ def check_args(parsed_args):
-        raise NotImplementedError('Backbone \'{}\' not implemented.'.format(parsed_args.backbone))
-
-    validate_backbone(parsed_args.backbone)
+    models.validate_backbone(parsed_args.backbone)
```
:+1:

de-vri-es commented on keras_retinanet/layers/filter_detections.py:
```diff
@@ -0,0 +1,198 @@
+        other           : List of tensors of shape (num_boxes, ...) to filter along with the boxes and classification scores.
+        nms             : Flag to enable/disable non maximum suppression.
+        score_threshold : Threshold used to prefilter the boxes with.
+        max_detections  : Maximum number of detections to filter.
```
It still says "filter" here.

de-vri-es commented on keras_retinanet/layers/filter_detections.py:
```diff
@@ -0,0 +1,198 @@
+            nms                 : Flag to enable/disable NMS.
+            nms_threshold       : Threshold for the IoU value to determine when a box should be suppressed.
+            score_threshold     : Threshold used to prefilter the boxes with.
+            max_detections      : Maximum number of detections to keep.
```
I see you did change it to "keep" here. :+1:

de-vri-es commented on keras_retinanet/models/__init__.py:
```diff
@@ -0,0 +1,58 @@
+def custom_objects(backbone):
```
It's good to factor out these operations from the scripts.

If we change this to objects per backbone with these operations it becomes a lot easier to add new backbones though. It would also be a good first step towards backbone plugins. Can be done in a future PR.

hgaiser replied:
I agree, let's do this in a different PR.

de-vri-es commented on keras_retinanet/models/densenet.py:
```diff
@@ -17,7 +17,7 @@
 from keras.applications.densenet import DenseNet, get_file
 
-from ..models import retinanet
+from . import retinanet
```
Nice catch.

de-vri-es commented on keras_retinanet/models/retinanet.py:
```diff
@@ -25,14 +25,14 @@
+    'UpsampleLike'     : layers.UpsampleLike,
+    'PriorProbability' : initializers.PriorProbability,
+    'RegressBoxes'     : layers.RegressBoxes,
+    'FilterDetections' : layers.FilterDetections,
```
Also for a future PR, but I think we should prefix out custom objects with `keras_retinanet.`, or maybe even use the full module path and class name. Otherwise it becomes hard (if not impossible) to mix retinanet with other models that use the same names in the same graph.

de-vri-es commented on keras_retinanet/models/retinanet.py:
```diff
@@ -330,42 +326,43 @@ def retinanet_bbox(
     """
-    model = retinanet(inputs=inputs, num_classes=num_classes, **kwargs)
+    if model is None:
+        model = retinanet(anchor_parameters=anchor_parameters, **kwargs)
```
How about just passing only `anchor_parameters.num_anchors()` to the `retinanet` function? The less coupling we have the better.

hgaiser replied:
:+1:

de-vri-es commented on keras_retinanet/utils/visualization.py:
```diff
@@ -59,41 +59,42 @@ def draw_boxes(image, boxes, color, thickness=2):
 
 
-def draw_annotations(image, annotations, color=(0, 255, 0), generator=None):
+def draw_annotations(image, annotations, color=(0, 255, 0), label_to_name=None):
```
:tada:

hgaiser commented:
Oh yeah, multi-gpu also seems to be working better after this PR. Whether it is working as intended is still to be determined (at the very least the outputs match with what happens without multi-gpu).
